### PR TITLE
Run check-commits dispatcher only if PR branch has `compile-commit` job

### PR DESCRIPTION
**Pull request by MiguelWeezardo**
## Description

#15157 introduced a composite job for compiling every commit of a PR.

Before running check-commits dispatcher it's important to check that the PR branch contains the definition of the `compile-commit` composite job.

A bug was also discovered in the job responsible for selecting commits to be compiled. 
The `rev-list` query by default returns the list of commit hashes in **reverse** chronological order.
The last `rev-list` entry removed by the `jq` pipeline expression was the oldest PR commit, not the newest one.

This commit fixes the problem by using the `rev-list` expression instead of `jq` pipeline to filter out PR HEAD.

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by MiguelWeezardo** [comment]
Tested with https://github.com/MiguelWeezardo/trino/pull/12

**Review by nineinchnick — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -99,10 +98,10 @@ jobs:
       - name: Set matrix
         id: set-matrix
         # The output from rev-list ends with a newline, so we have to filter out index -1 in jq since it's an empty string
-        # The HEAD commit of the PR (index -2) can be safely ignored since it's already compiled in other jobs
+        # The HEAD commit of the PR can be safely ignored since it's already compiled in other jobs
         run: |
-          export JQ_PIPELINE='split("\n") | .[0:-2] | map({base_ref: "${{ github.event.pull_request.base.ref }}", commit: .}) | select(length > 0) | {include: .}'
-          git rev-list refs/remotes/origin/${{ github.event.pull_request.base.ref }}..HEAD | jq --raw-input --slurp "$JQ_PIPELINE" > commit-matrix.json
+          export JQ_PIPELINE='split("\n") | .[0:-1] | map({base_ref: "${{ github.event.pull_request.base.ref }}", commit: .}) | select(length > 0) | {include: .}'
+          git rev-list --reverse refs/remotes/origin/${{ github.event.pull_request.base.ref }}..${{ github.event.pull_request.head.sha }}~ | jq --raw-input --slurp "$JQ_PIPELINE" > commit-matrix.json
```
Comment (nineinchnick) [inline .github/workflows/ci.yml:104]: What does the `~` do? Maybe add a comment

**Comment by MiguelWeezardo** [comment]
I've come to a conclusion that it's not possible to make this transition seamless for PRs created before this change was merged. It will fail in such cases, since the main workflow file will be taken from the temporary PR merge commit, but there will be no composite job definition on the PR branch being checked out in the dispatcher jobs.

**Comment by MiguelWeezardo** [comment]
During my investigation I have found a bug in the commit selection pipeline which I intend to fix here instead.

**Comment by MiguelWeezardo** [comment]
@hashhar Could you take a look at this and merge if it looks OK?

**Comment by hashhar** [comment]
will we actually have this problem now? This only affects PRs whose base commit is older than when the job was added?

I'm ok with adding this but I think at some point in future this will be redundant?

**Comment by MiguelWeezardo** [comment]
> will we actually have this problem now? This only affects PRs whose base commit is older than when the job was added?
> 
> I'm ok with adding this but I think at some point in future this will be redundant?

There are still open PRs which are based on older commits. Once these get updated with whatever review comments they had, their authors are in for a surprise when this check fails on them. This will lead to a lot of CI churn, since they will have to discover they need to rebase, and rerun the full CI pipeline after rebase. I'd like to avoid that if possible.

There's also another issue - while working on this I discovered and fixed a bug related to the order in which commits to compile were chosen. My intention was to compile all PR commits except the newest HEAD, but It turned out that I was compiling every commit except the oldest one.

**Review by hashhar — APPROVED** [review]

**Comment by hashhar** [comment]
Let's create an issue to "re-simplify" this at some point in future.

**Pull request merged**